In [8]:
import tensorflow as tf
old_v = tf.logging.get_verbosity()
tf.logging.set_verbosity(tf.logging.ERROR)

<h2>Extract MNIST data</h2>
<p style="font-size:20px">You can change the option of one_hot encoding.

In [2]:
from tensorflow.examples.tutorials.mnist import input_data
#get mnist data, with one_hot encoding
mnist = input_data.read_data_sets("MNIST_data/",one_hot=True)
#suppress warnings
tf.logging.set_verbosity(old_v)

Extracting MNIST_data/train-images-idx3-ubyte.gz
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz


<h2>Experiment</h2>
<p style="font-size:20px">

In [16]:
#learning rate
lr = 0.001
#number of traning steps
num_iteration = 2000
#number of batch_size
batch_size = 100

#network parameters
n_hidden_1 = 512
n_hidden_2 = 128
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.nn.relu(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.nn.relu(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_iteration):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
        if i % 1000 ==0:
            acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
            #print("step "+str(i)+", Accuracy= {:.3f}".format(acc))
    
    print("Training finished!")
    print("Activation_function: ReLU")
    print("num_iteration:" + str(num_iteration) +",batch_size:" + str(batch_size) +",n_hidden_1:"+ str(n_hidden_1)+",n_hidden_2:"+ str(n_hidden_2) +", learning_rate:" + str(lr))
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
Activation_function: ReLU
num_iteration:2000,batch_size:100,n_hidden_1:512,n_hidden_2:128, learning_rate:0.001
Testing Accuracy: 0.9163


In [37]:
#learning rate
lr = 0.01
#number of traning steps
num_iteration = 2000
#number of batch_size
batch_size = 100

#network parameters
n_hidden_1 = 512
n_hidden_2 = 128
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.nn.relu(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.nn.relu(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_iteration):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
        if i % 1000 ==0:
            acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
            #print("step "+str(i)+", Accuracy= {:.3f}".format(acc))
    
    print("Training finished!")
    print("Activation_function: ReLU")
    print("num_iteration:" + str(num_iteration) +",batch_size:" + str(batch_size) +",n_hidden_1:"+ str(n_hidden_1)+",n_hidden_2:"+ str(n_hidden_2) +", learning_rate:" + str(lr))
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
Activation_function: ReLU
num_iteration:2000,batch_size:100,n_hidden_1:512,n_hidden_2:128, learning_rate:0.01
Testing Accuracy: 0.8286


In [21]:
#learning rate
lr = 0.001
#number of traning steps
num_iteration = 2000
#number of batch_size
batch_size = 100

#network parameters
n_hidden_1 = 300
n_hidden_2 = 100
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.nn.relu(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.nn.relu(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_iteration):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
        if i % 1000 ==0:
            acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
            #print("step "+str(i)+", Accuracy= {:.3f}".format(acc))
    
    print("Training finished!")
    print("Activation_function: ReLU")
    print("num_iteration:" + str(num_iteration) +",batch_size:" + str(batch_size) +",n_hidden_1:"+ str(n_hidden_1)+",n_hidden_2:"+ str(n_hidden_2) +", learning_rate:" + str(lr))
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
Activation_function: ReLU
num_iteration:2000,batch_size:100,n_hidden_1:300,n_hidden_2:100, learning_rate:0.001
Testing Accuracy: 0.9084


In [28]:
#learning rate
lr = 0.001
#number of traning steps
num_iteration = 20000
#number of batch_size
batch_size = 100

#network parameters
n_hidden_1 = 512
n_hidden_2 = 128
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.nn.relu(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.nn.relu(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_iteration):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
        if i % 1000 ==0:
            acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
            #print("step "+str(i)+", Accuracy= {:.3f}".format(acc))
    
    print("Training finished!")
    print("Activation_function: ReLU")
    print("num_iteration:" + str(num_iteration) +",batch_size:" + str(batch_size) +",n_hidden_1:"+ str(n_hidden_1)+",n_hidden_2:"+ str(n_hidden_2) +", learning_rate:" + str(lr))
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
Activation_function: ReLU
num_iteration:20000,batch_size:100,n_hidden_1:512,n_hidden_2:128, learning_rate:0.001
Testing Accuracy: 0.9631


In [27]:
#learning rate
lr = 0.001
#number of traning steps
num_iteration = 20000
#number of batch_size
batch_size = 64

#network parameters
n_hidden_1 = 512
n_hidden_2 = 256
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.nn.relu(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.nn.relu(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_iteration):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
        if i % 1000 ==0:
            acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
            #print("step "+str(i)+", Accuracy= {:.3f}".format(acc))
    
    print("Training finished!")
    print("Activation_function: ReLU")
    print("num_iteration:" + str(num_iteration) +",batch_size:" + str(batch_size) +",n_hidden_1:"+ str(n_hidden_1)+",n_hidden_2:"+ str(n_hidden_2) +", learning_rate:" + str(lr))
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
Activation_function: ReLU
num_iteration:20000,batch_size:64,n_hidden_1:512,n_hidden_2:256, learning_rate:0.001
Testing Accuracy: 0.9654


In [31]:
#learning rate
lr = 0.01
#number of traning steps
num_iteration = 2000
#number of batch_size
batch_size = 100

#network parameters
n_hidden_1 = 512
n_hidden_2 = 128
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.sigmoid(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.sigmoid(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_iteration):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
        if i % 1000 ==0:
            acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
            #print("step "+str(i)+", Accuracy= {:.3f}".format(acc))
    
    print("Training finished!")
    print("Activation_function: Sigmoid")
    print("num_iteration:" + str(num_iteration) +",batch_size:" + str(batch_size) +",n_hidden_1:"+ str(n_hidden_1)+",n_hidden_2:"+ str(n_hidden_2) +", learning_rate:" + str(lr))
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
Activation_function: Sigmoid
num_iteration:2000,batch_size:100,n_hidden_1:512,n_hidden_2:128, learning_rate:0.01
Testing Accuracy: 0.9486


In [32]:
#learning rate
lr = 0.001
#number of traning steps
num_iteration = 2000
#number of batch_size
batch_size = 100

#network parameters
n_hidden_1 = 512
n_hidden_2 = 128
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.sigmoid(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.sigmoid(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_iteration):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
        if i % 1000 ==0:
            acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
            #print("step "+str(i)+", Accuracy= {:.3f}".format(acc))
    
    print("Training finished!")
    print("Activation_function: Sigmoid")
    print("num_iteration:" + str(num_iteration) +",batch_size:" + str(batch_size) +",n_hidden_1:"+ str(n_hidden_1)+",n_hidden_2:"+ str(n_hidden_2) +", learning_rate:" + str(lr))
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
Activation_function: Sigmoid
num_iteration:2000,batch_size:100,n_hidden_1:512,n_hidden_2:128, learning_rate:0.001
Testing Accuracy: 0.9062


In [33]:
#learning rate
lr = 0.001
#number of traning steps
num_iteration = 2000
#number of batch_size
batch_size = 100

#network parameters
n_hidden_1 = 300
n_hidden_2 = 100
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.sigmoid(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.sigmoid(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_iteration):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
        if i % 1000 ==0:
            acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
            #print("step "+str(i)+", Accuracy= {:.3f}".format(acc))
    
    print("Training finished!")
    print("Activation_function: Sigmoid")
    print("num_iteration:" + str(num_iteration) +",batch_size:" + str(batch_size) +",n_hidden_1:"+ str(n_hidden_1)+",n_hidden_2:"+ str(n_hidden_2) +", learning_rate:" + str(lr))
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
Activation_function: Sigmoid
num_iteration:2000,batch_size:100,n_hidden_1:300,n_hidden_2:100, learning_rate:0.001
Testing Accuracy: 0.9016


In [34]:
#learning rate
lr = 0.01
#number of traning steps
num_iteration = 20000
#number of batch_size
batch_size = 100

#network parameters
n_hidden_1 = 512
n_hidden_2 = 128
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.sigmoid(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.sigmoid(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_iteration):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
        if i % 1000 ==0:
            acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
            #print("step "+str(i)+", Accuracy= {:.3f}".format(acc))
    
    print("Training finished!")
    print("Activation_function: Sigmoid")
    print("num_iteration:" + str(num_iteration) +",batch_size:" + str(batch_size) +",n_hidden_1:"+ str(n_hidden_1)+",n_hidden_2:"+ str(n_hidden_2) +", learning_rate:" + str(lr))
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
Activation_function: Sigmoid
num_iteration:20000,batch_size:100,n_hidden_1:512,n_hidden_2:128, learning_rate:0.01
Testing Accuracy: 0.9676


In [36]:
#learning rate
lr = 0.001
#number of traning steps
num_iteration = 20000
#number of batch_size
batch_size = 100

#network parameters
n_hidden_1 = 512
n_hidden_2 = 128
num_input = 784
num_classes = 10

tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.sigmoid(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.sigmoid(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_iteration):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
        if i % 1000 ==0:
            acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
            #print("step "+str(i)+", Accuracy= {:.3f}".format(acc))
    
    print("Training finished!")
    print("Activation_function: Sigmoid")
    print("num_iteration:" + str(num_iteration) +",batch_size:" + str(batch_size) +",n_hidden_1:"+ str(n_hidden_1)+",n_hidden_2:"+ str(n_hidden_2) +", learning_rate:" + str(lr))
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

Training finished!
Activation_function: Sigmoid
num_iteration:20000,batch_size:100,n_hidden_1:512,n_hidden_2:128, learning_rate:0.001
Testing Accuracy: 0.9562


<h2>Define hyperparameters</h2>

In [38]:
#learning rate
lr = 0.001
#number of traning steps
num_iteration = 20000
#number of batch_size
batch_size = 100

#network parameters
n_hidden_1 = 512
n_hidden_2 = 256
num_input = 784
num_classes = 10

<h2>Define placeholder and Variables</h2>

In [39]:
tf.reset_default_graph()
#tf graph input
X = tf.placeholder(tf.float32,[None,num_input],name='X')
Y = tf.placeholder(tf.int32,[None,num_classes],name='Y')

#Layers weight & bias
weights = {
    'W1': tf.Variable(tf.random_normal([num_input, n_hidden_1]),name='W1'),
    'W2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2]),name='W2'),
    'Wout': tf.Variable(tf.random_normal([n_hidden_2, num_classes]),name='Wout')
}

biases = {
    'b1': tf.Variable(tf.zeros(shape=[n_hidden_1]),name='b1'),
    'b2': tf.Variable(tf.zeros(shape=[n_hidden_2]),name='b2'),
    'bout': tf.Variable(tf.zeros(shape=[num_classes]),name='bout')
}

<h2>Define neural network</h2>

In [40]:
def neural_net(x):
    layer_1_out = tf.add(tf.matmul(x,weights['W1']),biases['b1'])
    layer_1_out = tf.nn.relu(layer_1_out)
    layer_2_out = tf.add(tf.matmul(layer_1_out,weights['W2']),biases['b2'])
    layer_2_out = tf.nn.relu(layer_2_out)
    out = tf.add(tf.matmul(layer_2_out,weights['Wout']),biases['bout'])
    return out

<h2>Define cost function and accuracy</h2>

In [41]:
#predicted labels
logits = neural_net(X)

#define loss
loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(logits=logits,labels=Y),name='loss')
#define optimizer
optimizer = tf.train.AdamOptimizer(learning_rate=lr)
train_op = optimizer.minimize(loss)

#compare the predicted labels with true labels
correct_pred = tf.equal(tf.argmax(logits,1),tf.argmax(Y,1))

#compute the accuracy by taking average
accuracy = tf.reduce_mean(tf.cast(correct_pred,tf.float32),name='accuracy')

#Initialize the variables
init = tf.global_variables_initializer()

<h2>Execute training</h2>

In [42]:
with tf.Session() as sess:
    sess.run(init)
    
    for i in range(num_iteration):
        #fetch batch
        batch_x, batch_y = mnist.train.next_batch(batch_size)
        #run optimization
        sess.run(train_op, feed_dict={X:batch_x, Y:batch_y})
        if i % 1000 ==0:
            acc = sess.run(accuracy,feed_dict={X:batch_x, Y:batch_y})
            print("step "+str(i)+", Accuracy= {:.3f}".format(acc))
    
    print("Training finished!")
    
    print("Testing Accuracy:", sess.run(accuracy, feed_dict={X:mnist.test.images, Y:mnist.test.labels}))

step 0, Accuracy= 0.220
step 1000, Accuracy= 0.940
step 2000, Accuracy= 0.940
step 3000, Accuracy= 0.940
step 4000, Accuracy= 0.990
step 5000, Accuracy= 1.000
step 6000, Accuracy= 0.990
step 7000, Accuracy= 1.000
step 8000, Accuracy= 0.990
step 9000, Accuracy= 0.990
step 10000, Accuracy= 1.000
step 11000, Accuracy= 1.000
step 12000, Accuracy= 1.000
step 13000, Accuracy= 1.000
step 14000, Accuracy= 0.990
step 15000, Accuracy= 1.000
step 16000, Accuracy= 1.000
step 17000, Accuracy= 1.000
step 18000, Accuracy= 1.000
step 19000, Accuracy= 1.000
Training finished!
Testing Accuracy: 0.9639


<h2>Your results</h2>

| Activation_function | num_iteration | batch_size | n_hidden_1 | n_hidden_2 | learning_rate | accuracy |
|---------------------|---------------|------------|------------|------------|---------------|----------|
| Sigmoid             | 2000          | 100        | 300        | 100        | 0.001         | 0.9016   |
| Sigmoid             | 2000          | 100        | 512        | 128        | 0.01          | 0.9486   |
| Sigmoid             | 2000          | 100        | 512        | 128        | 0.001         | 0.9062   |
| Sigmoid             | 20000         | 100        | 512        | 128        | 0.01          | 0.9676   |
| Sigmoid             | 2000          | 100        | 512        | 128        | 0.001         | 0.9562   |
| ReLU                | 2000          | 100        | 300        | 100        | 0.001         | 0.9084   |
| ReLU                | 2000          | 100        | 512        | 128        | 0.01          | 0.8286   |
| ReLU                | 2000          | 100        | 512        | 128        | 0.001         | 0.9163   |
| ReLU                | 20000         | 64         | 512        | 128        | 0.001         | 0.9654   |
| ReLU                | 20000         | 100        | 512        | 128        | 0.001         | 0.9631   |